File: scala/src/main/scala/zalando/ToString.scala
package zalando

object ToString {
  def main(args: Array[String]): Unit = {
    println(
      toBeautifulString(
        List(
          List("one", "two", "three"),
          List("four", "five"),
          List("six", "seven"),
          List("eight", "nine", "ten", "eleven")
        )
      )
    )
  }


  /*
    input
    [
        ["one", "two", "three"],
        ["four", "five"],
        ["six"]
    ]

    output
    | one  | two  | three |
    | four | five |
    | six  |
 */
  def toBeautifulString(xs: List[List[String]]): String = {
    val maxLengths = maxLengthByColumn(xs)

    xs.map(
      // ["one", "two"] -> | one  | two  |
      x => x.zip(maxLengths).map {
        case (s, l) => String.format("%-" + l + "s", s)
      }.mkString("| ", " | ", " |")
    ).mkString("\n")
  }

  /*
    input
    [
        ["one", "two", "three"],
        ["four", "five"],
        ["six"]
    ]
    output
         4       4       5
 */
  def maxLengthByColumn(xs: List[List[String]]): List[Int] =
    xs.foldLeft(List[Int]())(
      (x1, x2) => x1.zipAll(x2, Int.MinValue, Int.MinValue)
        .map {
          case (v, s: String) => v max s.length
          case (v, s: Int) => v max s
        }
    )

  //      map(_.map(_.length))
  //      .reduceLeft(
  //        (x1, x2) =>
  //          x1.zipAll(x2, Int.MinValue, Int.MinValue)
  //            .map { case (a, b) => a max b }
  //      )

}

File: java-kotlin/src/zalando/ToString.java
package zalando;

import java.util.ArrayList;
import java.util.List;

public class ToString {

    public static void main(String[] args) {
        System.out.println(
                new ToString().toString(
                        List.of(
                            List.of("one", "two", "three"),
                            List.of("four", "five"),
                            List.of("six")
                        )
                )
        );
    }

    /*
        input
        [
            ["one", "two", "three"],
            ["four", "five"],
            ["six"]
        ]

        output
        | one  | two  | three |
        | four | five |
        | six  |
     */
    public String toString(List<List<String>> input) {

        final List<Integer> longestLengthByColumn = new ArrayList<>();
        for (List<String> row : input) {
            for (int i = 0; i < row.size(); i++) {
                final int length = row.get(i).length();
                try {
                    longestLengthByColumn.set(i, Math.max(longestLengthByColumn.get(i), length));
                } catch (IndexOutOfBoundsException e) {
                    longestLengthByColumn.add(length);
                }
            }
        }

        final StringBuilder sb = new StringBuilder();
        for (List<String> row : input) {
            sb.append("|");
            for (int i = 0; i < row.size(); i++) {
                sb.append(String.format(" %-" + longestLengthByColumn.get(i) + "s |", row.get(i)));
            }
            sb.append("\n");
        }

        return sb.toString();
    }
}